src/future/index.js

```diff
@@ -0,0 +1,108 @@
+import React, { useState, useMemo } from 'react';
+
+import useField from '@/src/future/hooks/useField';
+
+import { FieldContext, RenderContext } from '@/src/future/context';
+
+import Component from '@/src/future/components/Component';
+import Field from '@/src/future/components/Field';
+
+import TextInput from '@/src/future/inputs/TextInput';
+
+function ControlledFormContainer({ value, onChange, children }) {
+  function onSetFieldValue(fieldKey, fieldValue) {
+    const newValue = {
+      ...value,
+      [fieldKey]: fieldValue,
+    };
+    onChange(newValue);
+  }
+  return (
+    <FieldContext.Provider value={{ formValue: value, onSetFieldValue }}>
+      {children}
+    </FieldContext.Provider>
+  );
+}
+
+function UncontrolledFormContainer({ defaultValue, onChange, children }) {
+  const [formValue, setFormValue] = useState(defaultValue);
+
+  function onChangeControlled(newValue) {
+    setFormValue(newValue);
+    onChange(newValue);
+  }
+
+  return (
+    <ControlledFormContainer onChange={onChangeControlled} value={formValue}>
+      {children}
+    </ControlledFormContainer>
+  );
+}
+
+export function FormContainer({
+  value,
+  defaultValue,
+  renderTag,
+  renderComponent,
+  ...props
+}) {
+  const [savedDefaultValue] = useState(defaultValue);
+  const [isControlled] = useState(savedDefaultValue === undefined);
+  // Keep context object pure.
+  const renderContext = useMemo(() => ({ renderTag, renderComponent }), [
+    renderTag,
+    renderComponent,
+  ]);
+  return (
+    <RenderContext.Provider value={renderContext}>
+      {isControlled ? (
+        <ControlledFormContainer value={value} {...props} />
+      ) : (
+        <UncontrolledFormContainer
+          defaultValue={savedDefaultValue}
+          {...props}
+        />
+      )}
+    </RenderContext.Provider>
+  );
+}
+
+export function FieldContainer({ fieldKey, children, ...props }) {
+  const { value, onChange, onChangeTargetValue } = useField(fieldKey);
+  if (typeof children === 'function') {
+    return children({
+      fieldKey,
+      value,
+      onChangeTargetValue,
+      ...props,
+    });
+  }
+  return (
+    <ControlledFormContainer onChange={onChange} value={value}>
+      {children}
+    </ControlledFormContainer>
+  );
+}
+
+function createField(fieldType, Input) {
+  function FieldComponent(props) {
+    return (
+      <Component
+        _component={Field}
```

Comment:
Is the intent for `_component` and other underscored props to imply do not mutate? Why wouldn't `fieldType` be `_fieldType`?

I've only really used `_foo` with object fields and methods—not as much with function args.

Reply:
The underscored props were just to signify that they would not be passed on to the component/tag. If a component had its own `component` prop, then you could still use it like:

```js<Component